packages/compass-import-export/src/utils/collection-stream.spec.ts

```diff
@@ -0,0 +1,280 @@
+import type { Writable } from 'stream';
+import { expect } from 'chai';
+
+import { createCollectionWriteStream } from './collection-stream';
+
+const BATCH_SIZE = 1000;
+
+function getDataService({
+  isFLE,
+  throwErrors,
+}: {
+  isFLE: boolean;
+  throwErrors: boolean;
+}) {
+  return {
+    bulkWrite: (ns: string, docs: any[] /*, options: any*/) => {
+      return new Promise((resolve, reject) => {
+        if (isFLE && docs.length !== 1) {
+          const error: any = new Error(
+            'Only single insert batches are supported in FLE2'
+          );
+          error.code = 6371202;
+          return reject(error);
+        }
+
+        if (throwErrors) {
+          const error = new Error('fake bulkWrite error');
+          delete error.stack; // slows down tests due to excess output
+          return reject(error);
+        }
+
+        resolve({
+          nInserted: docs.length,
+          nMatched: 0,
+          nModified: 0,
+          nRemoved: 0,
+          nUpserted: 0,
+          ok: 1,
+        });
+      });
+    },
+
+    insertOne: (
+      ns: string,
+      doc: any,
+      options: any,
+      callback: (err: Error | null, result?: any) => void
+    ) => {
+      if (throwErrors) {
+        const error = new Error('fake insertOne error');
+        delete error.stack; // slows down tests due to excess output
+        callback(error);
+        return;
+      }
+
+      callback(null, { acknowledged: true });
+    },
+  };
+}
+
+async function insertDocs(dest: Writable, docs: any) {
+  try {
+    for (const doc of docs) {
+      await new Promise<void>((resolve, reject) => {
+        dest.write(doc, (err) => (err ? reject(err) : resolve()));
+      });
+    }
+
+    return new Promise<void>((resolve, reject) => {
+      dest.end((err?: Error) => (err ? reject(err) : resolve()));
+    });
+  } catch (err) {
+    // we'll get here if stopOnErrors is true, because dest.write will throw
+    // ignore this for now (and stop writing more). we'll detect it via the
+    // stream's error event in the tests
+  }
+}
+
+function getExpectedNumBatches(
+  numDocs: number,
+  isFLE: boolean,
+  stopOnErrors: boolean
+) {
+  if (stopOnErrors) {
+    return 1;
+  }
+
+  if (isFLE) {
+    // one attempted batch at the batch size (followed by insertOne() on retry), then subsequent batches are all size 1.
+    return numDocs > BATCH_SIZE ? 1 + numDocs - BATCH_SIZE : 1;
```

Comment:
☝️ this shows how often the progress event can be called for FLE2. And that becomes a call straight to dispatch() here: https://github.com/mongodb-js/compass/blob/6f0050b373c609355b053aa3c643c494b6177eb5/packages/compass-import-export/src/modules/import.ts#L290-L292